### PR TITLE
Feature/add variable groups support

### PR DESCRIPTION
**Pull request by ignatz42**
# PR Summary

Added functions to Add, Get, Update, and Remove Variable groups.

## PR Checklist

- [x] [Write Help](https://github.com/DarqueWarrior/vsteam/blob/master/.github/CONTRIBUTING.md#write-help)
- [x] [Write Unit Test](https://github.com/DarqueWarrior/vsteam/blob/master/.github/CONTRIBUTING.md#write-unit-test)
- [x] [Update VSTeam.psd1](https://github.com/DarqueWarrior/vsteam/blob/master/.github/CONTRIBUTING.md#add-a-format-file)
- [x] [Update CHANGELOG.md](https://github.com/DarqueWarrior/vsteam/blob/master/.github/CONTRIBUTING.md#add-a-format-file)

### CODE REVIEW

**Comment by ignatz42** [comment]
Feature request: https://github.com/DarqueWarrior/vsteam/issues/24

**Comment by DarqueWarrior** [comment]
Reviewing now.

File: .docs/Update-VSTeamVariableGroup.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+Update-VSTeamVariableGroup [-id] <String> [-variableGroupName] <String> [-variableGroupType] <String>
+ [-variableGroupDescription] <String> [-variableGroupVariables] <Hashtable>
+ [[-variableGroupProviderData] <Hashtable>] [-Force] [-WhatIf] [-Confirm] [-ProjectName] <String>
+ [<CommonParameters>]
+```
+
+## DESCRIPTION
+<!-- #include "./synopsis/Update-VSTeamVariableGroup.md" -->
+
+## EXAMPLES
```
Comment (DarqueWarrior) [inline .docs/Update-VSTeamVariableGroup.md]: Please provide example.

File: .docs/Get-VSTeamVariableGroup.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+### ByID
+```powershell
+Get-VSTeamVariableGroup -id <String> [-ProjectName] <String> [<CommonParameters>]
+```
+
+## DESCRIPTION
+<!-- #include "./synopsis/Get-VSTeamVariableGroup.md" -->
+## EXAMPLES
+
+### Example 1
```
Comment (DarqueWarrior) [inline .docs/Get-VSTeamVariableGroup.md]: You have the heading for an example but you did not provide one.

File: .docs/Add-VSTeamVariableGroup.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
+<!-- #include "./common/header.md" -->
+
+# Add-VSTeamVariableGroup
+
+## SYNOPSIS
+<!-- #include "./synopsis/Add-VSTeamServiceEndpoint.md" -->
+
+## SYNTAX
+
```
Comment (DarqueWarrior) [inline .docs/Add-VSTeamVariableGroup.md]: You can leave the syntax area blank. When the help is shown it is determined from the attributes of the parameters.  Reviewing the help will let you know if you have all the attributes of the parameters are correct.

File: .docs/Add-VSTeamVariableGroup.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+## PARAMETERS
+
+### -ProjectName
+The name of the project. 
+You can tab complete from the projects in your Team Services or TFS account when passed on the command line.
+
+```yaml
+Type: String
+Parameter Sets: (All)
```
Comment (DarqueWarrior) [inline .docs/Add-VSTeamVariableGroup.md]: You don't need to add (All) to each parameter when you only have a default parameter set for the function.

File: integration/test/010_projects.Tests.ps1 [hunk 5]
```diff
@@ -309,6 +309,39 @@ InModuleScope VSTeam {
          }
       }
 
+      Context 'Simple Variable Group' {
```
Comment (DarqueWarrior) [inline integration/test/010_projects.Tests.ps1]: Add tests for get list and get by id.

**Comment by ignatz42** [comment]
Thanks for the considered review!  I think I've addressed the file comments. I need a bit of time to sort out the unit tests.  I don't have TFS available to test against but I'll try to follow the existing tests.

**Comment by ignatz42** [comment]
I've added the unit tests along with the other requested changes. Will you please review the updates?

**Comment by DarqueWarrior** [comment]
> Thanks for the considered review! I think I've addressed the file comments. I need a bit of time to sort out the unit tests. I don't have TFS available to test against but I'll try to follow the existing tests.

I will send you a DM and give you access to my TFS 2017 and 2018 test machines.

**Comment by DarqueWarrior** [comment]
Playing with my TFS2017 I found that variable groups are supported with version 3.2-preview.1. I will update your PR for you.

**Comment by DarqueWarrior** [comment]
TFS2017 supports variable groups with version 3.2-preview.1. I added to Set-VSTeamAPIVersion but I think the shape of the JSON might be different. I am willing to give you access to my 2017 and 2018. You can email me if you are not on twitter. I am dbrown at microsoft.

**Comment by ignatz42** [comment]
I sent you an email earlier in the week but I suspect it was caught in a spam filter so I've just now sent another with the title "VSTeam PullRequest 170".  Meanwhile, I've updated the unit tests to use JSON samples that I collected using Invoke-VSTSRequet calls.  If possible would you please add a similar sample from VSTS 2017?  If so, I'll update the unit test to use the new file and then fix any failing code as necessary.

**Comment by ignatz42** [comment]
Thanks for the access to the test system.  I think I've addressed the TFS2017 differences with this last round of commits. I'll look forward to the next review.

**Comment by DarqueWarrior** [comment]
> Thanks for the access to the test system. I think I've addressed the TFS2017 differences with this last round of commits. I'll look forward to the next review.

Did you run the integration tests?  None of them pass for the variable groups. If you open 000_team.Tests.ps1 file it will tell you which environment variables to set to run the tests. I will email you a Azure DevOps organization you can use to run these tests. DO NOT RUN on your own accounts as they delete everything!  Expect email soon.

**Comment by ignatz42** [comment]
Thanks for your email with instructions on performing the integration tests, much time was saved.  The integration tests on all three target types complete successfully now.

**Comment by DarqueWarrior** [comment]
Well done! Your code is on it way to the PowerShell Gallery right now. Are you on Twitter? If so I can tag you in the announcement of the new version.

**Comment by ignatz42** [comment]
Thanks you very much! I'm not on Twitter but I am excited to put the changes to work.

**Pull request merged**